athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchMetadataHandler.java

```diff
@@ -265,43 +274,59 @@ public void getPartitions(BlockWriter blockWriter, GetTableLayoutRequest request
-     * 1. A Set<Split> which represent read operations Amazon Athena must perform by calling your read function.
+     * 1. A Set<Split> each containing a domain and endpoint, and the shard to be retrieved by the Record handler.
      * 2. (Optional) A continuation token which allows you to paginate the generation of splits for large queries.
-     * @note A Split is a mostly opaque object to Amazon Athena. Amazon Athena will use the optional SpillLocation and
-     * optional EncryptionKey for pipelined reads but all properties you set on the Split are passed to your read
-     * function to help you perform the read.
-     * @throws RuntimeException when the domain does not exist in the map.
+     * @throws RuntimeException when the domain does not exist in the map, or an error occurs while processing the
+     * cluster/shard health information.
      */
     @Override
     public GetSplitsResponse doGetSplits(BlockAllocator allocator, GetSplitsRequest request)
             throws RuntimeException
     {
         logger.debug("doGetSplits: enter - " + request);
 
+        // Create set of splits
+        Set<Split> splits = new LinkedHashSet<>();
+        // Get domain
         String domain = request.getTableName().getSchemaName();
-
-        // Every split must have a unique location if we wish to spill to avoid failures
-        SpillLocation spillLocation = makeSpillLocation(request);
-
-        // Create split
-        Split.Builder splitBuilder = Split.newBuilder(spillLocation, makeEncryptionKey());
+        // Get index
+        String index = request.getTableName().getTableName();
 
         try {
             String endpoint = getDomainEndpoint(domain);
-            // Add domain and endpoint to the split to be used by the Record Handler.
-            splitBuilder.add(domain, endpoint);
+            AwsRestHighLevelClient client = clientFactory.getClient(endpoint);
+            try {
+                Map<Integer, ClusterShardHealth> shardHealthInfo = client.getShardHealthInfo(index);
+                for (Map.Entry<Integer, ClusterShardHealth> entry : shardHealthInfo.entrySet()) {
+                    Integer shardId = entry.getKey();
+                    ClusterShardHealth shardHealth = entry.getValue();
+                    // Process only primary active shards.
+                    if (shardHealth.isPrimaryActive()) {
+                        // Every split must have a unique location if we wish to spill to avoid failures
+                        SpillLocation spillLocation = makeSpillLocation(request);
+                        // Create a new split (added to the splits set) that includes the domain and endpoint, and
+                        // shard information (to be used later by the Record Handler).
+                        splits.add(new Split(spillLocation, makeEncryptionKey(), ImmutableMap
+                                .of(domain, endpoint, SHARD_KEY, SHARD_VALUE + shardId.toString())));
+                    }
```

Comment:
What does it mean if this is false? should we fail the query? will we potentially miss data if we proceed without it?

Reply:
I use primary shards as a differentiator from replicated shards do avoid possible duplication. Shard health can be unpredictably in flux. A shard can be currently active, or in some unavailable state. Rather then wait for all shards to be active or fail the request, I think we should strive for eventual consistency and work with what we have. The assumption here is that some of the data is better than none. I do fail the order if none of the shards are active.

Reply:
I dont think we can allow eventual consistency here. We need some way to know if every share has at least 1 active replica in the elastic-search cluster or we should fail the query. Is that possible?

Reply:
It seems that the shard-health query returns only one set of shards for the index without duplicates. Removed check for primary shards.